File: apps/flutter-app/lib/core/services/user_service.dart
import 'dart:math';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';

import '../constants/firestore_collections.dart';
import '../models/user_model.dart';

class UserService {
  const UserService(this._firestore);

  final FirebaseFirestore _firestore;

  Future<void> createUser(UserModel user) async {
    await _firestore
        .collection(FirestoreCollections.users)
        .doc(user.uid)
        .set(user.toFirestore());
    await _firestore.waitForPendingWrites();
  }

  Future<UserModel?> getUser(String uid) async {
    debugPrint('[UserService] getUser: $uid');
    final doc =
        await _firestore.collection(FirestoreCollections.users).doc(uid).get();
    if (!doc.exists) return null;
    return UserModel.fromFirestore(doc);
  }

  Future<List<UserModel>> getVolunteers() async {
    final snapshot = await _firestore
        .collection(FirestoreCollections.users)
        .where('role', isEqualTo: UserRole.volunteer.value)
        .get();
    return snapshot.docs.map(UserModel.fromFirestore).toList();
  }

  Future<UserModel?> getRandomVolunteer() async {
    final volunteers = await getVolunteers();
    if (volunteers.isEmpty) return null;
    return volunteers[Random().nextInt(volunteers.length)];
  }

  Future<List<UserModel>> getStudentsForVolunteer(String volunteerId) async {
    final snapshot = await _firestore
        .collection(FirestoreCollections.users)
        .where('role', isEqualTo: UserRole.student.value)
        .where('volunteer_id', isEqualTo: volunteerId)
        .get();
    return snapshot.docs.map(UserModel.fromFirestore).toList();
  }
}

File: apps/angular-app/src/app/core/services/user.service.ts
import { inject, Injectable } from '@angular/core';
import { collection, doc, getDoc, setDoc, Timestamp } from 'firebase/firestore';
import { FIREBASE_FIRESTORE } from '../tokens/firebase.tokens';
import { FIRESTORE_COLLECTIONS } from '../constants/firestore-collections';
import type { User } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly firestore = inject(FIREBASE_FIRESTORE);

  async getUser(uid: string): Promise<User | null> {
    const ref = doc(collection(this.firestore, FIRESTORE_COLLECTIONS.USERS), uid);
    const snap = await getDoc(ref);
    if (!snap.exists()) return null;
    const data = snap.data();
    return {
      uid: snap.id,
      firstName: data['first_name'] as string,
      lastName: data['last_name'] as string,
      nickname: data['nickname'] as string,
      email: data['email'] as string,
      birthdate: (data['birthdate'] as Timestamp).toDate(),
      role: data['role'] as 'student' | 'volunteer',
      createdAt: (data['created_at'] as Timestamp).toDate(),
      updatedAt: (data['updated_at'] as Timestamp).toDate(),
    };
  }

  async createUser(user: User): Promise<void> {
    const ref = doc(collection(this.firestore, FIRESTORE_COLLECTIONS.USERS), user.uid);
    await setDoc(ref, {
      uid: user.uid,
      first_name: user.firstName,
      last_name: user.lastName,
      nickname: user.nickname,
      email: user.email,
      birthdate: Timestamp.fromDate(user.birthdate),
      role: user.role,
      created_at: Timestamp.fromDate(user.createdAt),
      updated_at: Timestamp.fromDate(user.updatedAt),
    });
  }
}
